src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,397 @@
+  inline arma::Cube<eT> CubeMultiply(const arma::Cube<eT>& a,
+    const arma::Cube<eT>& b,
+    const bool aTranspose = false,
+    const bool bTranspose = false)
+  {
+    size_t rows = a.n_rows, cols = b.n_cols, slices = a.n_slices;
+    Log::Assert(a.n_slices == b.n_slices);
+    if (bTranspose)
+    {
+      Log::Assert(a.n_cols == b.n_cols);
+      cols = b.n_rows;
+    }
+    else if (aTranspose)
+    {
+      Log::Assert(a.n_rows == b.n_cols);
+      rows = a.n_cols;
+    }
+    else
+      Log::Assert(a.n_cols == b.n_rows);
+
+    arma::Cube<eT> z(rows, cols, slices);
+
+    if (bTranspose)
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a.slice(i) * b.slice(i).t();
+    }
+    else if (aTranspose)
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a.slice(i).t() * b.slice(i);
+    }
+    else
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a.slice(i) * b.slice(i);
+    }
+    return z;
+  }
+
+  //! Function to multiply a matrix object to a cube object.
+  template <typename eT = double>
+  inline arma::Cube<eT> CubeMultiply(const arma::Mat<eT>& a,
+    const arma::Cube<eT>& b,
+    const bool aTranspose = false,
+    const bool bTranspose = false)
+  {
+    size_t rows = a.n_rows, cols = b.n_cols, slices = b.n_slices;
+    if (bTranspose)
+    {
+      Log::Assert(a.n_cols == b.n_cols);
+      cols = b.n_rows;
+    }
+    else if (aTranspose)
+    {
+      Log::Assert(a.n_rows == b.n_cols);
+      rows = a.n_cols;
+    }
+    else
+      Log::Assert(a.n_cols == b.n_rows);
+
+    arma::Cube<eT> z(rows, cols, slices);
+
+    if (bTranspose)
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a * b.slice(i).t();
+    }
+    else if (aTranspose)
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a.t() * b.slice(i);
+    }
+    else
+    {
+      for (size_t i = 0; i < slices; ++i)
+        z.slice(i) = a * b.slice(i);
+    }
+    return z;
+  }
+
+  template <typename eT>
+  void Expand(const arma::Mat<eT>& query,
+              arma::Mat<eT>& key,
+              arma::Mat<eT>& value,
+              arma::Cube<eT>& q,
+              arma::Cube<eT>& k,
+              arma::Cube<eT>& v)
+  {
+    typedef typename arma::Cube<eT> CubeType;
+
+    Log::Assert(query.n_rows % embedDim == 0, "Number of features in 'query' \
+        must be divisible by embedding dimension");
+    Log::Assert(key.n_rows % embedDim == 0, "Number of features in 'key' \
+        must be divisible by embedding dimension");
+    Log::Assert(arma::size(key) == arma::size(value), "'key' and 'value' \
+        must have same dimensions.");
+
+    q.set_size(embedDim, query.n_rows / embedDim, query.n_cols);
+
+    for (size_t i = 0; i < query.n_elem; ++i)
+    {
+      const size_t row = i % embedDim;
+      const size_t col = (i / embedDim) % (query.n_rows / embedDim);
+      const size_t slice = i / query.n_rows;
+      q(row, col, slice) = query(i);
+    }
```

Comment:
Do you suggest
```cpp
q = CubeType(const_cast<arma::Mat<eT>&>(query).memptr(),
        embedDim, query.n_rows / embedDim, query.n_cols, false, false);
```
?
I think this removes the constantness of `query`, and later `q` is modified also. I preferred copying it (and turning from 2D to 3D).

Even if we set `copy_aux_mem = true`, I think we will still be removing constantness of `query`(which is the actual input) and copying also. What should I do?

Reply:
No, `copy_aux_mem` equals `true` by default. So, I suggest
```c++
q = CubeType(const_cast<arma::Mat<eT>&>(query).memptr(),
        embedDim, query.n_rows / embedDim, query.n_cols, true, false);
```
Yes, it will remove constantness of `query`, but it won't modify `query`. As far as I know it's not forbidden.